docs/exernal-services.md

```diff
@@ -0,0 +1,206 @@
+
+Additionally if the user is a teacher then the teacher object is added to the access grant.  If the user is a student then the learner object (student, offering) is added to the access grant. These additions to the grant are used when the report requests a JWT. The issued JWT then includes claims based on the teacher or learner. References:
+- https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/controllers/api/v1/jwt_controller.rb#L59
+- https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/controllers/api/v1/jwt_controller.rb#L89
+
+# Researcher Reports
+
+**TODO**
+
+# SSO clients
+
+**TODO**
+
+# From a Modeling Point of View
+
+## Tool
+In the portal there is a tools model. This model has a tool_id (url), source_type, and name. An ExternalActivity can be associated with a single Tool.
+
+The source_type of the ExternalActivity#tool is used to find an ExternalReport configured with a default_report_for_source_type.
+
+In the user researcher report filer page, only ExternalActivites that have a Tool with a source_type of "LARA" are shown in the "runnables" list.
+
+When the (current defunct) move student feature is used, the portal sends the tool_id (url) of the ExternalActivities in the original class and new class of the student, to the ExternalReports#move_students_api_url
+
+The tool_id is used when an external activity is created using the `api/v1/external_activity` api.
+
+When ActivityRuntimeAPI is used to publish a new activity or sequence a new Tool is created if the source_type is passed in. In this case the Tool has no tool_id. This feature is mainly for developer convenience.
+
+## ExternalReport
+Has a `default_report_for_source_type`. When an ExternalActivity has a tool with this source_type is assigned to a class, then report buttons for this assignment will automatically be added.
+
+ExternalActivities can also have 1 or more ExternalReports directly associated with them. This makes it possible for an author or admin to setup a special report or dashboard for a specific resource.
+
+## ExternalActivity
+Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the a special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as a 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.
```

Comment:
```suggestion
Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as an 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.
```